libgalois/include/katana/PropertyGraph.h

```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
+    CreateLabeledAdjIndex();
+  }
+
+  const PropertyGraph* prop_graph_;
+  const EdgeLabelIndex* edge_label_index_;
+  AdjIndexArray adj_indices_;
+  EdgeDstArray dests_;
+  OrigEdgeIDMap orig_edge_ids_;
+};
+
+/**
+ * Provides both out-going and in-coming topology API
+ * Provides Label Aware topology API
+ */
+class KATANA_EXPORT EdgeLabelAwareBiDirTopology
+    : public EdgeLabelAwareTopology {
+  // Inheriting to reuse outgoing topology API
+  //
+  using OutTopoBase = EdgeLabelAwareTopology;
+
+  using OutTopoBase::EdgeLabelTy;
+
+public:
+  static EdgeLabelAwareBiDirTopology Make(const PropertyGraph* pg) noexcept {
+    // EdgeLabelIndex edge_label_index = EdgeLabelIndex::Make(pg);
+    // OutTopoBase out_topo = OutTopoBase::MakeFromDefaultTopology(pg, &edge_label_index);
+    // EdgeLabelAwareTopology in_topo = EdgeLabelAwareTopology::MakeFromTransposeTopology(pg, &edge_label_index);
+    // return EdgeLabelAwareBiDirTopology{std::move(edge_label_index), std::move(out_topo), std::move(in_topo)};
+
+    auto edge_label_index =
+        std::make_unique<EdgeLabelIndex>(EdgeLabelIndex::Make(pg));
+    return EdgeLabelAwareBiDirTopology{pg, std::move(edge_label_index)};
+  }
+
+  EdgeLabelAwareBiDirTopology() = default;
+  EdgeLabelAwareBiDirTopology(EdgeLabelAwareBiDirTopology&&) = default;
+  EdgeLabelAwareBiDirTopology& operator=(EdgeLabelAwareBiDirTopology&&) =
+      default;
+
+  EdgeLabelAwareBiDirTopology(const EdgeLabelAwareBiDirTopology&) = delete;
+  EdgeLabelAwareBiDirTopology& operator=(const EdgeLabelAwareBiDirTopology&) =
+      delete;
+
+  bool has_edge_label(const EdgeLabelTy& label) const noexcept {
+    return edge_label_index_->has_edge_label(label);
+  }
+
+  auto distinct_edge_labels() const noexcept {
+    return edge_label_index_->distinct_edge_labels();
+  }
+
+  edge_iterator in_edge_begin(Node N, const EdgeLabelTy& label) const noexcept {
+    return in_topo_.edge_begin(N, label);
+  }
+
+  edge_iterator in_edge_end(Node N, const EdgeLabelTy& label) const noexcept {
+    return in_topo_.edge_end(N, label);
+  }
+
+  edge_iterator in_edge_begin(Node N) const noexcept {
+    return in_topo_.edge_begin(N);
+  }
+
+  edge_iterator in_edge_end(Node N) const noexcept {
+    return in_topo_.edge_end(N);
+  }
+
+  edges_range in_edges(Node N) const noexcept { return in_topo_.edges(N); }
+
+  edges_range in_edges(Node N, const EdgeLabelTy& label) const noexcept {
+    return in_topo_.edges(N, label);
+  }
+
+  Node in_edge_dest(Edge edge_id) const noexcept {
+    return in_topo_.edge_dest(edge_id);
+  }
+
+  Node in_edge_dest(const edge_iterator ei) const noexcept {
+    return in_topo_.edge_dest(ei);
+  }
+  size_t in_degree(Node N) const noexcept { return in_topo_.degree(N); }
+
+  size_t in_degree(Node N, const EdgeLabelTy& label) const noexcept {
+    return in_topo_.degree(N, label);
+  }
+
+  Edge original_edge_id_using_in_edge(
+      const edge_iterator& in_edge_iter) const noexcept {
```

Comment:
I'd like to get ride of the edge_iterator variants of these if possible. Having two versions just doesn't make much sense. And the iterator version is the one people use when they are NOT using range-for and I want to encourage people to use range-for.

Reply:
Can do.

Reply:
done.